stdlib/public/core/StringGutsRangeReplaceable.swift

```diff
@@ -94,20 +94,28 @@ extension _StringGuts {
     if let currentCap = self.uniqueNativeCapacity, currentCap >= n { return }
 
     // Grow
-    self.grow(n) // TODO: no factor should be applied
+    self.grow(n)
   }
 
   // Grow to accommodate at least `n` code units
   @usableFromInline
   internal mutating func grow(_ n: Int) {
-    defer { self._invariantCheck() }
+    defer {
+      self._invariantCheck()
+      _internalInvariant(
+        self.uniqueNativeCapacity != nil && self.uniqueNativeCapacity! >= n)
+    }
 
     _internalInvariant(
       self.uniqueNativeCapacity == nil || self.uniqueNativeCapacity! < n)
 
-    // TODO: Don't do this! Growth should only happen for append...
-    let growthTarget = Swift.max(n, (self.uniqueNativeCapacity ?? 0) * 2)
+    // We're non-unique OR smaller than requested, so as we copy to new storage,
+    // choose a size that at least stores the current contents.
+    let growthTarget = Swift.max(n, self.utf8Count)
```

Comment:
I was trying to decide if in the case of unique native storage whether should we also lower-bound this by the current capacity. But since we’re definitely creating a copy at this point, that doesn’t seem necessary. So this LGTM.

Reply:
That was my initial thought, but we actually [have a test codifying the current behavior](https://github.com/apple/swift/blob/main/test/stdlib/NewStringAppending.swift#L136-L148), which shrinks the new storage to fit the size of the contents. Didn't want to end up with unexpectedly large capacities, so now this better matches what we're doing now.